FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -27,16 +27,48 @@ public final class CoreDataFeedStore: FeedStore {
 		)
 		context = container.newBackgroundContext()
 	}
+}
 
+extension CoreDataFeedStore: FeedStore {
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let managedCache = try ManagedCache.newUniqueInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				try ManagedCache.deleteCache(in: context)
+				try context.save()
```

Comment:
In case there is no previously stored cache `context.delete` will not change the context, thus there will be no changes in it and will be no need to save it. Otherwise, Core Data may perform unnecessary work.

One way to ensure the context is only saved when there are changes is to use `context.hasChanges`:
```suggestion
				try ManagedCache.deleteCache(in: context)
				if context.hasChanges {
					try context.save()
				}
```